### PR TITLE
Add solution for izho-17-LBS

### PR DESCRIPTION
**Pull request by rsj001**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

File: solutions/izho-17-LongestBeautifulSequence.mdx [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+id: izho-17-LongestBeautifulSequence
+source: IZhO 2017
+title: 2017 - Longest beautiful sequence
+author: Shijie Ren
+---
+
+<Spoiler title="Hint 1">
+
+Both $\mathcal{O}(N^2)$ and $\mathcal{O}(N\max{a[i]})$ algorithms are not hard to
+
```
Comment (envyaims) [inline solutions/izho-17-LongestBeautifulSequence.mdx:11]: why are the spacings kind of weird?

File: solutions/izho-17-LongestBeautifulSequence.mdx [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int maxcount = 10;
+const int mask = 1 << maxcount;
+const int maxn = 100005;
+
+int cnt[mask];  // popcount array
```
Comment (envyaims) [inline solutions/izho-17-LongestBeautifulSequence.mdx:148]: From: https://usaco.guide/general/adding-solution?lang=cpp

image

File: solutions/izho-17-LongestBeautifulSequence.mdx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+and query are unbalanced.
+
+Try to consider the two equations carefully and try to "combine" their ideas.
+
+
+</Spoiler>
+
+<Spoiler title="Explanation">
+
+$\mathcal{O}(N^2)$ DP:
```
Comment (envyaims) [inline solutions/izho-17-LongestBeautifulSequence.mdx:41]: maybe make these complexity notations bold so they stand out more?

File: solutions/izho-17-LongestBeautifulSequence.mdx [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+$$
+\texttt{dp}[i][j][q]=\texttt{dp}[i-1][j][q],q\not=\text{popcount}(a[i]\&j)
+$$
+
+According to Hint 2, we need to consider either of them, that is, to consider
+the answers before and after.
+
+Since $a[i] < 2^{20}$ , divide the 20-bits number into two 10-bits numbers.
+
+For convenience, let $\text{b0}(x) = \lfloor \frac{x}{2^{10}} \rfloor$
```
Comment (envyaims) [inline solutions/izho-17-LongestBeautifulSequence.mdx:93]: i feel like better naming than "b0" and "b1" would be helpful

File: solutions/izho-17-LongestBeautifulSequence.mdx [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+
+Let $\texttt{ans}[i]$ be the length of longest beautiful sequence that ends with
+$a[i]$, and $\texttt{dp}[p][i][j][q]$ be the length of longest beautiful
+sequence of the first $p$ elements which satisfies $i = \text{b1}(\texttt{EndElement})$,
+
+$
+\text{popcount}(\text{b0}(\texttt{EndElement})\&j)=q$.
+
+For convenience, $\texttt{cur0} = \text{b0}(a[i])$,
```
Comment (envyaims) [inline solutions/izho-17-LongestBeautifulSequence.mdx:104]: these names aren't that necessary

File: solutions/izho-17-LongestBeautifulSequence.mdx [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+		cin >> a[i];
+	}
+	for (int i = 1; i <= n; i++) {
+		cin >> k[i];
+	}
+	int maxid = 0, maxn = 0;
+	for (int i = 1; i <= n; i++) {
+		int cur0 = (a[i]) >> maxcount, cur1 = (a[i]) & (mask - 1), ans = 1;
+		for (int pre1 = 0; pre1 < mask; pre1++) {
+			int bc = k[i] - cnt[pre1 & cur1];
```
Comment (envyaims) [inline solutions/izho-17-LongestBeautifulSequence.mdx:173]: comments throughout here would be nice

**Review by envyaims — CHANGES_REQUESTED** [review]
the spacing in the non-code parts are kind of weird

also generally more comments would be nice so the reader doesn't get lost

File: solutions/izho-17-LongestBeautifulSequence.mdx (hunk 2 above)
Comment (rsj001) [reply solutions/izho-17-LongestBeautifulSequence.mdx:148]: but it's constant

File: solutions/izho-17-LongestBeautifulSequence.mdx (hunk 3 above)
Comment (rsj001) [reply solutions/izho-17-LongestBeautifulSequence.mdx:41]: ok

File: solutions/izho-17-LongestBeautifulSequence.mdx (hunk 6 above)
Comment (rsj001) [reply solutions/izho-17-LongestBeautifulSequence.mdx:173]: ok i'll fix later

**Pull request closed without merging**